### ISSUE TITLE
Extending burnup and cycle inputs

### ISSUE THREAD

**Issue by keckler** [terrapower/armi#576]
Building off of the discussion in #561, a number of upgrades/extensions to the current cycle input are planned. This ticket will keep track of the various pieces.

## Purpose
Right now the structure of a burnup case is defined by a few case settings: `nCycles`, `power`, `powerFraction(s)`, `burnSteps`, `availabilityFactor(s)`, and `cycleLength(s)`. These settings allows one to define a case with multiple cycles, where each cycle can have variable length, power, and availability. However it has a number of drawbacks that are not ideal for modeling some realistic systems:
1. Power within the cycle cannot vary, so if a user wants to change the power level they must define an entirely new cycle
2. The length of each burnup step is fixed to be uniform within each cycle, so capturing some important physics like Xenon buildup can be difficult

This ticket will primarily address these two shortcomings, in addition to a handful of associated enhancements.
## Needs
- [x] Allow for variable numbers of burn steps within each cycle
- [x] Allow for non-uniform burn step lengths within each cycle
- [x] Allow for power to change at each burn step within each cycle

## Would-be-nice-to-have's
- [x] Allow burn step lengths to be input as either cumulatives or deltas
- [ ] Allow powers to be input as either fractions or absolutes
- [x] Allow for cycles to have names
- [ ] Allow for the decay period currently defined by `availabilityFactor(s)` to contain multiple steps
- [x] Continue to allow the use of repeated lists as is currently allowed within vectors like `availabilityFactors` and `cycleLengths`
- [ ] Only trigger a specific subset of interfaces on zero-power cycles (i.e. decay only)

## Method
These capabilities will be enabled for the user through the use of new yaml-style input options to be implemented in the case settings file. This input should notionally look something like the following:

```
cycles:
  - name: "Cycle 1 (initial power tests)"
    cumulative days: [1, 5, 10, 100]
    powerFractions: [0.1, 0.5, 1, 1]
  - name: "Cycle 2.a" 
    cumulative days: [1, 5, 6, 15, 50]
    powerFractions: [0.1, 0.5, 0.7, 1.0, 0.9]
  - name: "Cycle 2.b" 
    cumulative days: [50]
    powerFractions: [1.0]
```

These changes should maintain backwards compatibility by retaining the current cycle input in addition to the new method. The user will be able to use either format, and the code will resolve whichever input format that the user decides to use into the same data internal structure. 

## Other considerations
Allowing for the burn steps and power to be variable will require some changes throughout the code base and related plugins. Right now `burnSteps` is stored in the case settings and is accessed via calls like `self.cs["burnSteps"]`. But since `burnSteps` will now depend on the current cycle, the input will need to be resolved and stored on the operator. This is likely to break plugins. So all calls throughout the codebase and plugins will need to be changed to call to a new method which pulls the needed data off the operator.

**Comment by keckler** [comment]
Note that the changes in #637 now allow for the cycles to have names, but nothing is being done with the names once they are resolved. They are stored on the operator, so they can be accessed throughout a run, but nothing in the framework is using the names yet.

**Comment by keckler** [comment]
Based on my digging, I think it will be somewhat complicated to enable the "would-be-nice-to-have"
>  Allow for the decay period currently defined by availabilityFactor(s) to contain multiple steps

So abandoning that for now.

**Comment by keckler** [comment]
Closed with the closure of #637 . Note that some of the "would-be-nice-to-have"s didn't get implemented. That's okay for now.
